x-pack/plugin/geoip-enterprise-downloader/src/main/java/org/elasticsearch/xpack/geoip/EnterpriseGeoIpDownloaderLicenseListener.java

```diff
@@ -74,47 +82,54 @@ void listenForLicenseStateChanges() {
         licenseState.addListener(this);
     }
 
+    @NotMultiProjectCapable(description = "Replace DEFAULT project after enterprise license is supported in serverless and project-aware")
     @Override
     public void licenseStateChanged() {
-        licenseIsValid = ENTERPRISE_GEOIP_FEATURE.checkWithoutTracking(licenseState);
-        maybeUpdateTaskState(clusterService.state());
+        licenseIsValid.put(ProjectId.DEFAULT, ENTERPRISE_GEOIP_FEATURE.checkWithoutTracking(licenseState));
+        maybeUpdateTaskState(clusterService.state().projectState(ProjectId.DEFAULT));
     }
 
     @Override
     public void clusterChanged(ClusterChangedEvent event) {
-        hasIngestGeoIpMetadata = event.state().metadata().getProject().custom(INGEST_GEOIP_CUSTOM_METADATA_TYPE) != null;
-        final boolean ingestGeoIpCustomMetaChangedInEvent = event.metadataChanged()
-            && event.changedCustomProjectMetadataSet().contains(INGEST_GEOIP_CUSTOM_METADATA_TYPE);
-        final boolean masterNodeChanged = Objects.equals(
-            event.state().nodes().getMasterNode(),
-            event.previousState().nodes().getMasterNode()
-        ) == false;
-        /*
-         * We don't want to potentially start the task on every cluster state change, so only maybeUpdateTaskState if this cluster change
-         * event involved the modification of custom geoip metadata OR a master node change
-         */
-        if (ingestGeoIpCustomMetaChangedInEvent || (masterNodeChanged && hasIngestGeoIpMetadata)) {
-            maybeUpdateTaskState(event.state());
-        }
+        event.state().forEachProject(projectState -> {
+            ProjectId projectId = projectState.projectId();
+            final boolean hasMetadata = event.state().metadata().getProject(projectId).custom(INGEST_GEOIP_CUSTOM_METADATA_TYPE) != null;
+            hasIngestGeoIpMetadata.put(projectId, hasMetadata);
+            final boolean ingestGeoIpCustomMetaChangedInEvent = event.metadataChanged()
+                && event.customMetadataChanged(projectId, INGEST_GEOIP_CUSTOM_METADATA_TYPE);
+            final boolean masterNodeChanged = Objects.equals(
+                event.state().nodes().getMasterNode(),
+                event.previousState().nodes().getMasterNode()
+            ) == false;
+            /*
+             * We don't want to potentially start the task on every cluster state change, so only maybeUpdateTaskState
+             * if this cluster change event involved the modification of custom geoip metadata OR a master node change
+             */
+            if (ingestGeoIpCustomMetaChangedInEvent || (masterNodeChanged && hasIngestGeoIpMetadata.getOrDefault(projectId, false))) {
+                maybeUpdateTaskState(projectState);
```

Comment:
We could also pass `event.state().nodes().isLocalNodeElectedMaster()` as a parameter together with a project ID. That way we don't need to convert the projects into project states. What do you think?